src/mscorlib/src/System/Threading/ThreadPool.cs

```diff
@@ -1625,7 +1675,23 @@ private static bool QueueUserWorkItemHelper(WaitCallback callBack, Object state,
                 try { }
                 finally
                 {
-                    QueueUserWorkItemCallback tpcallBack = new QueueUserWorkItemCallback(callBack, state, compressStack, ref stackMark);
+                    ExecutionContext context = compressStack && !ExecutionContext.IsFlowSuppressed() ?
+                        ExecutionContext.Capture(ref stackMark, ExecutionContext.CaptureOptions.IgnoreSyncCtx | ExecutionContext.CaptureOptions.OptimizeDefaultCase) :
+                        null;
+
+                    IThreadPoolWorkItem tpcallBack;
+                    if (context == null)
+                    {
+                        tpcallBack = new QueueUserWorkItemCallback(callBack, state, null); // null context
+                    }
+                    else if (context.IsDefaultFTContext(true))
+                    {
+                        tpcallBack = new QueueUserWorkItemCallbackDefaultContext(callBack, state); // default context
+                    }
+                    else
+                    {
+                        tpcallBack = new QueueUserWorkItemCallback(callBack, state, context); // non-default context
+                    }
```

Comment:
We don't need three cases here.  It can just be two:
```C#
IThreadPoolWorkItem tpcallBack = context != null && context.IsDefaultFTContext(true) ?
    new QueueUserWorkItemCallbackDefaultContext(callBack, state) :
    new QueueUserWorkItemCallback(callBack, state, context);
```